compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -121,15 +124,30 @@ impl<Tag> Allocation<Tag> {
         Allocation::from_bytes(slice, Align::ONE, Mutability::Not)
     }
 
-    pub fn uninit(size: Size, align: Align) -> Self {
-        Allocation {
-            bytes: vec![0; size.bytes_usize()],
+    /// Try to create an Allocation of `size` bytes, failing if there is not enough memory
+    /// available to the compiler to do so.
+    pub fn uninit(size: Size, align: Align) -> InterpResult<'static, Self> {
```

Comment:
Concretely, what I imagine is having a `bool` parameter here that indicates whether we should `try_reserve` or not. In `memory.rs` that `bool` parameter would be se by a `Machine` associated const. Then ConstProp can set that to `false`, and we need not worry about `MemoryExhausted` in `ConstProp`.